File: app/Models/StudyFocusParticipation.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class StudyFocusParticipation extends Model
{
    use HasFactory;

    public const STATUS_ACTIVE = 'active';

    public const STATUS_CANCELLED = 'cancelled';

    public const STATUS_COMPLETED = 'completed';

    protected $fillable = [
        'study_focus_room_id',
        'study_session_id',
        'user_id',
        'study_subject_id',
        'started_at',
        'ended_at',
        'paused_at',
        'paused_seconds',
        'duration_seconds',
        'status',
    ];

    protected function casts(): array
    {
        return [
            'started_at' => 'datetime',
            'ended_at' => 'datetime',
            'paused_at' => 'datetime',
            'paused_seconds' => 'integer',
            'duration_seconds' => 'integer',
        ];
    }

    public function focusRoom(): BelongsTo
    {
        return $this->belongsTo(StudyFocusRoom::class, 'study_focus_room_id');
    }

    public function studySession(): BelongsTo
    {
        return $this->belongsTo(StudySession::class);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function studySubject(): BelongsTo
    {
        return $this->belongsTo(StudySubject::class);
    }

    public function getDurationLabelAttribute(): string
    {
        return gmdate('H:i:s', (int) ($this->duration_seconds ?? 0));
    }

    public function effectiveElapsedSeconds(): int
    {
        $end = $this->ended_at ?: ($this->paused_at ?: now());
        $pausedSeconds = (int) ($this->paused_seconds ?? 0);

        return max(0, (int) round($this->started_at->diffInSeconds($end)) - $pausedSeconds);
    }

    public function isPaused(): bool
    {
        return $this->status === self::STATUS_ACTIVE && $this->paused_at !== null;
    }
}

File: app/models/study_focus_participation.rb
class StudyFocusParticipation < ApplicationRecord
  STATUS_ACTIVE = "active"
  STATUS_CANCELLED = "cancelled"
  STATUS_COMPLETED = "completed"

  belongs_to :study_focus_room
  belongs_to :study_session, optional: true
  belongs_to :user
  belongs_to :study_subject

  scope :active, -> { where(status: STATUS_ACTIVE) }

  def duration_label
    Time.at(duration_seconds.to_i).utc.strftime("%H:%M:%S")
  end

  # Focus-room participation mirrors the same elapsed-time logic as study
  # sessions so the public presence JSON and the timer widget stay in sync.
  def effective_elapsed_seconds
    end_time = ended_at || paused_at || Time.current
    paused_total = paused_seconds.to_i
    [ (end_time.to_i - started_at.to_i) - paused_total, 0 ].max
  end

  def paused?
    status == STATUS_ACTIVE && paused_at.present?
  end
end
